### ISSUE TITLE
[Feature Request] Supersample to Native Resolution

### ISSUE THREAD

**Issue by amiraizat** [PCSX2/pcsx2#4576]
It would be great if we can supersample games in high resolution (1080p), then downsample and display it back in native resolution (480p) and preserve that resolution even when in full screen.

From my experience games always look the best, or at least closest to the developer's intended look when played at native resolution, but playing fullscreen at native resolution with modern monitors comes with severe aliasing. Being able to supersample to native resolution fixes the aliasing issue (there will still be aliasing but it's much more pleasant to look at).

RetroArch's Beetle PSX HW have this feature. Duckstation has something similar too, but uses SSAA.

**Comment by seta-san** [comment]
That might be a cool mode to play in

**Comment by OCRBonk** [comment]
Yeah, the current resolution scaling doesn't work all that well. As the higher the ratio gets from display resolution the worse it tends to fare. In the included shaders currently are some options for Bicubic/Lanczos filtering but as far as I can tell this just applies those as a full screen filter prior to downsampling. Not as an actual downsampling  resolve function. Those would provide a much better result than the current solution.  If you have an Nvidia card you can use DSR, set that as your desktop resolution and then run PCSX2 to get an idea of how much better it could look with a better resolve.

It won't get all the aliasing (Texture aliasing is a big problem) but it would be an improvement without having to use a workaround.

**Comment by NecroPoster** [comment]
100% would love this feature. The overly sharp 3D edges of having an higher internal resolution alongside the low resolution 2D assets has always bugged me to no end. Downsampling (or the closest thing in terms to it SSAA) would completely solve this issue.

**Comment by refractionpcsx2** [comment]
Okay just to quell any excitement.  Doing SSAA or MSAA type equivalents is pretty much a non-starter.  The problem is we can't anti-alias things that are being drawn because many games rely on the accurate pixels being drawn otherwise post processing effects get completely broken.

The only option we have is to do post processing (like FXAA and SMAA), we might be able to combine thing with FSR to get *something* slightly better, however if you're hoping for amazing upscaling from it, don't get excited, again it needs a better AA pass than post processing, so a post effect "might" get a mild upscale effect out of it, it won't get much, the biggest thing you'll get out of it is a nice sharpening filter.

TL;DR Probably not happening nor possible, but we might be able to implement a post processing AA effect (without any real upscaling)

**Comment by NecroPoster** [comment]
Just curious, wouldn't it be possible to reduce a higher internal res image back down to native after the post processing effects have been drawn on screen.

**Comment by refractionpcsx2** [comment]
> Just curious, wouldn't it be possible to reduce a higher internal res image back down to native after the post processing effects have been drawn on screen.

in software mode? no, not efficiently. but also the act of drawing a higher resolution image will break the emulation and software mode is intended to be a reference/accurate version.  Plus the whole idea with things like FSR is you render at a lower than current resolution then use fancy techniques to upscale the image to a higher resolution, thus reducing the load on the computer.  Rendering at a higher resolution and downscaling would do the opposite, which is kinda pointless.

**Comment by NecroPoster** [comment]
> 
> in software mode? no, not efficiently. but also the act of drawing a higher resolution image will break the emulation and software mode is intended to be a reference/accurate version. Plus the whole idea with things like FSR is you render at a lower than current resolution then use fancy techniques to upscale the image to a higher resolution, thus reducing the load on the computer. Rendering at a higher resolution and downscaling would do the opposite, which is kinda pointless.
 
First off, I wasn't talking about software mode. Why would you say it's pointless tho? The whole point of downsample the way I see it is so we can have a better looking native res (as in PS2 res not monitor res. So I don't think FSR would matter here.) image with proper "anti-aliasing" on 3D edges but without having the ugly mismatch of having really high res 3D against low res 2D elements. So the "pixels" on the downsampled 3D edges looks the same as the ones in the 2D background. A picture of what I mean from Duckstation.

![comparison](https://user-images.githubusercontent.com/31898926/132241918-820ef7ba-9d13-46de-8047-3955635760fe.png)

**Comment by RedDevilus** [comment]
It's integer scaling what you want? It should help with 2D elements. As PCSX2 does have supersample effects on 3D by doing quite some upscaling above monitor resolution but not really on 2D.

**Comment by NecroPoster** [comment]
> 
> 
> It's integer scaling what you want? It should help with 2D elements. As PCSX2 does have supersample effects on 3D by doing quite some upscaling above monitor resolution but not really on 2D.

Not really interger scaling. Simply put it would be supersampling but instead of monitor res it would be to the game's native res.

**Comment by refractionpcsx2** [comment]
> I wasn't talking about software mode. Why would you say it's pointless tho? The whole point of downsample the way I see it is so we can have a better looking native res (as in PS2 res not monitor res. So I don't think FSR would matter here.) image with proper "anti-aliasing" on 3D edges but without having the ugly mismatch of having really high res 3D against low res 2D elements. So the "pixels" on the downsampled 3D edges looks the same as the ones in the 2D background. A picture of what I mean from Duckstation.
> 

the point is why would you want a better looking native when you can just have the upscaled version? what do you hope to achieve by downscaling it then attempting to upscale it with post processing effects all you're going to do is lose quality, it's not going to emulate any better (since you've wasted time rendering an upscaled version)?

I can see the scaled down version looks better, sure, but why? what benefit is there when you can just render in the higher resolution?

**Comment by tadanokojin** [comment]
I don't know if it was mentioned on Github yet, but I've said I'm willing to implement upscaling and downsampling back to native. It's simple enough to do, just use the upscaled framebuffer and do an extra draw into a 1x framebuffer at the end. Just gonna take time, we're in the middle of a bunch of rewrites right now and I need to redo our frontend presentation code.

**Comment by OCRBonk** [comment]
All PCSX2 really needs is improved resampling support for HW mode. No matter what internal resolution you use, the scaling to display res (Whether that's 640x480 or 1920x1200) is not optimal.
As nice as it would be to have MSAA (because on Nvidia cards you could then piggy back off that to use SGSSAA and super sample pixel shading per sub sample and that would fix 99% of the texture aliasing present in PS2 games.)

Something strange (I am not a coder) happens with the way the viewport is stretched to the right aspect ratio and then resampled to display/window size at high internal resolutions. The option "Texture Filtering of Display" seems to be responsible for this, otherwise you get scaling artifact and pixel soup. But it also doesn't seem to be doing something right, as even just taking the screenshot and doing a simple 1 step Bilinear Stretch and resample provides better results. 

At 1080p the 4x resolution is about the threshold of no return before the resolve starts falling apart and creating additional aliasing that didn't previously exist. In some cases there are very large gaps in edge gradients so big it's almost as if pixels are missing.

_Here are some comparisons:_
**5x Res**
https://www.screenshotcomparison.com/comparison/17185
**7x res**
I didn't crop this one because it's just so much more obvious, the higher the internal resolution the worse it looks.
https://www.screenshotcomparison.com/comparison/17186

**Comment by refractionpcsx2** [comment]
I'm sure kojin could explain in more detail but, that's kind of not our fault most of the time, a lot of games downsample the frame buffer and redraw it over itself to do kind of lighting effects n stuff, and some of these tricks completely destroy upscaling.

I don't think anything will be as good as proper upscaling, but we're always looking for cool tech we can maybe use

**Comment by OCRBonk** [comment]
Not asking for upscaling. Just asking for proper downsampling of the internal resolution to improve Anti Aliasing.
There's no point running a game at a higher internal resolution when the program can't internally handle the resolve of the final frame buffer from the input resolution. 

In the case of those 7x res examples [this is the source image](https://u.cubeupload.com/MrBonk/gs20210906190413.jpg) and the only difference is how you get the final 1080p image from this frame. PCSX2 is already doing something here to correct the aspect ratio and resample it to the display window resolution right? Wouldn't that step be the only thing to be changed?
(Or in the case of those people who want it downsampled to another resolution, couldn't there just be an option like the "Texture Filtering of Display" like "Downsample resolution to [insert user specified resolution here]" so someone who wanted to play the game at 5x res downsampled to 720x480 and then upscaled to their display could do so? (I personally don't see the benefit of this unless you were playing on a low res display to begin with. Which would make such an option irrelevant anyway)

**Comment by seta-san** [comment]
won't the upscaling artifacts still happen and be visible in the downscale?

**Comment by kozarovv** [comment]
Actually I'm curious about same thing as seta. Doesn't something like that bring all current upscaling bugs to native/downscaled resolution then?

**Comment by refractionpcsx2** [comment]
yes, the downscaled picture will suffer from all the bugs and problems introduced in upscaling the picture, so you will need the upscaling hacks to counter them in order to keep your picture looking clean (if they work for said game)

**Comment by RedDevilus** [comment]
I believe this fixed it: https://github.com/PCSX2/pcsx2/issues/5387

**Comment by MattHorn25** [comment]
Implementing downsampling sounds like a great idea - and Duckstation indeed already has it implemented for a reference on how it can be done. Is there maybe any kind of movement on this front? @tadanokojin mentioned being interested in implementing this? I can vouch for this as a feature that a subset of users (me included) would be very happy to have.

**Comment by stenzek** [comment]
Bumping threads/nagging people isn't going to make it happen any faster. In fact, quite often it just does the opposite - if it's not something we're interested in doing ourselves, being annoying further reduces the motivation. You're always welcome to implement it yourself if it matters so much to you.
